module/VuFindSearch/src/VuFindSearch/Backend/Solr/Command/WriteDocumentCommand.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+/**
+ * Command to write a document object to Solr.
+ *
+ * PHP version 7
+ *
+ * Copyright (C) Villanova University 2021.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Search
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org
+ */
+namespace VuFindSearch\Backend\Solr\Command;
+
+use VuFindSearch\Backend\BackendInterface;
+use VuFindSearch\Backend\Solr\Connector;
+use VuFindSearch\Backend\Solr\Document\DocumentInterface;
+use VuFindSearch\Command\CommandInterface;
+use VuFindSearch\ParamBag;
+
+/**
+ * Command to write a document object to Solr.
+ *
+ * @category VuFind
+ * @package  Search
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org
+ */
+class WriteDocumentCommand extends \VuFindSearch\Command\AbstractBase
+{
+    /**
+     * Constructor.
+     *
+     * @param string            $backend Search backend identifier
+     * @param DocumentInterface $doc     Document to write
+     * @param ?int              $timeout Timeout value (null for default)
+     * @param string            $handler Handler to use
+     * @param ?ParamBag         $params  Search backend parameters
+     */
+    public function __construct(
+        string $backend,
+        DocumentInterface $doc,
+        int $timeout = null,
+        string $handler = 'update',
+        ?ParamBag $params = null
+    ) {
+        parent::__construct($backend, compact('doc', 'timeout', 'handler'), $params);
+    }
+
+    /**
+     * Execute command on backend.
+     *
+     * @param BackendInterface $backend Backend
+     *
+     * @return CommandInterface Command instance for method chaining
+     */
+    public function execute(BackendInterface $backend): CommandInterface
+    {
```

Comment:
Could this code not go into a `Backend::writeDocument()` method?

Reply:
True, I'll refactor.

Reply:
Okay, this is done in 56766dc (and f330f42, since I inadvertently broke some tests). I think it's much better now! Do you mind taking one last look? If you're satisfied now, I think this refactoring project is complete (unless there's more to do with #1886 before the 8.0 release).

Reply:
(And I see now that you're proposing we wait on #1886, so that's one less thing to worry about this week :-) ).